### ISSUE TITLE
Upgrade wget.exe application

### ISSUE THREAD

**Issue by moteus** [luarocks/luarocks#713]
Today when I try install my lua-cURL library I get error 
```
>luarocks install lua-curl --server=https://luarocks.org/dev
Installing https://luarocks.org/dev/lua-curl-scm-0.rockspec

Error: Error fetching file: Failed downloading https://github.com/Lua-cURL/Lua-cURLv3/archive/master.zip - master.zip
```
I found that problem with wget. When I run wget for this url I got
```
--2017-06-29 19:59:24--  https://github.com/Lua-cURL/Lua-cURLv3/archive/master.zip
Resolving github.com... 192.30.253.112, 192.30.253.113
Connecting to github.com|192.30.253.112|:443... connected.
Unable to establish SSL connection.
```
I think problem that wget 1.11.4 does not support TLSv1.2
So after I update wget.exe to 1.19.1 I be able install it without any problems,

*PS.* Also may be switch to curl?

**Comment by Tieske** [comment]
@moteus would you mind sending a pr with the updated binary?

**Comment by hishamhm** [comment]
I found a binary here: https://eternallybored.org/misc/wget/

However, the binary is 3.4MB in size, compared to the current 439KB wget.exe we currently ship.

```
hisham@packet master ~/luarocks/win32/tools]file wget.exe
wget.exe: PE32 executable (console) Intel 80386 (stripped to external PDB), for MS Windows
hisham@packet master ~/luarocks/win32/tools]file wget.exe.1 
wget.exe.1: PE32 executable (console) Intel 80386, for MS Windows
```

Looks like our current binary is stripped, but the other one isn't. Can anyone with access to Windows strip the binary? I found this tool that seems to do the job: http://www.jrsoftware.org/striprlc.php

**Comment by moteus** [comment]
I suspect that this wget just statically link with all deps. I think most of size is iconv library.
I try tool you mentioned. It did nothing.

**Comment by hishamhm** [comment]
@moteus Thank you. Do you have a smaller wget.exe binary for the latest version that we could use? 

Alternatively, can we remove the DLLs from `win32/tools` if we ship this `wget.exe`?  Currently, our `tools` directory includes:

* libeay32.dll - 1,177,600
* libiconv2.dll - 1,008,128
* libintl3.dll - 103,424
* libssl32.dll - 232,960

I don't know if the other exe files there need any of these libs. Can anyone verify?

**Comment by moteus** [comment]
> Do you have a smaller wget.exe binary for the latest version that we could use?
No. I use the same binary.

>Alternatively, can we remove the DLLs from win32/tools if we ship this wget.exe?
`libintl3.dll` depend on `libiconv2.dll`. and almost all exe files depend on `libintl3.dll`.
But no one depend on openssl. (At least statically).

And one more. Just test old wget and it now works again. I am be able download this file without any
problem.
So may be we can wait for now :)

**Comment by hishamhm** [comment]
> Hrm, is the shipped libeay32.dll used by lua libraries (e.g. luaossl?)

I don't think so, IIRC it's there because of wget.exe only.

**Comment by blandcr** [comment]
Hi,

As of yesterday (22/02/2018) github appears to have (officially) dropped support for TLS 1.0 and 1.1[1]. This change doesn't seem to have completely propagated across github infrastructure as of today but already some packages are failing to install (suchas `inspect 3.1.0-1` and the latest Moses, though weirdly `inspect 3.1.1-0` installed fine earlier.)

[1] https://githubengineering.com/crypto-removal-notice/

If you need some help making a reasonably sized build of `wget` please let me know what the requirements are and I can look in to solve this. Thanks.

**Comment by ildar** [comment]
Isn't it a good idea to get rid of binaries in repo and store them
somewhere else? E.g. in release packs.
I hate when sources are bloated with binary garbage.

**Comment by hishamhm** [comment]
> If you need some help making a reasonably sized build of wget please let me know what the requirements are and I can look in to solve this. Thanks.

@blandcr yes, please! I have received requests to switch from wget to curl as well, so that's perhaps a better option (curl is already the default on the Mac).

> Isn't it a good idea to get rid of binaries in repo and store them somewhere else? E.g. in release packs.

@ildar it would, but we would also need somewhere to keep them, so the script can fetch them when packaging the releases.

Perhaps the ideal situation would be to set up a repo that builds all necessary binaries in Appveyor, creates a LuaRocks release package and uploads it somewhere. Would that be doable, I wonder?

**Comment by blandcr** [comment]
@hishamhm OK. I do also think curl is probably a better option since it seems to have more consistent updates for windows binaries. I can put up a PR w/ a switch to default using curl 7.58 later tonight and we can go from there!

**Comment by moteus** [comment]
For curl there also few choices.
What type of crypto library use. OpenSSL/SSPI/some other
If use OpenSSL do we need use static linking or
use dynamic. And if using dynamic openssl do 
we need distribute it  as well and were put it on destination system?

**Comment by blandcr** [comment]
@moteus For curl I was planning on using the `curl.exe` from the build here: [https://skanthak.homepage.t-online.de/curl.html](url) (it is linked from the official curl site as the latest build for win32 w/ SSL but not ssh. I picked this since we probably care about binary size.)

This uses WinSSL/SSPI? I am not super knowledgeable about SSL libraries.

The following is the result of `dumpbin /dependents curl.exe` from that executable:
```
Microsoft (R) COFF/PE Dumper Version 14.11.25547.0
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file curl.exe

File Type: EXECUTABLE IMAGE

  Image has the following dependencies:

    WS2_32.dll
    CRYPT32.dll
    ADVAPI32.dll
    WLDAP32.dll
    KERNEL32.dll

  Summary

        4000 .data
       4B000 .rdata
        5000 .rsrc
       6B000 .text
```

So I think as far as dependencies go this is probably fine unless it needs to work on pre-win2k. I do not know about any sort of licensing issues but presumably building against Microsoft APIs is fine?

**Comment by hishamhm** [comment]
@blandcr this binary looks like a good option! @moteus's build could work as well.

> Fix the hardcoded wget usage, remove wget entirely and switch to curl by default. This is more involved but probably the right way to go about doing stuff.

Agreed!

> I do think a decision should be made on (2) but in the mean time it is possible to resolve the test_env.lua and cache.lua hardcoded dependency on wget.

Yes, we can eventually move them out of the main repo but that can be done as a separate step, beyond the scope of fixing the wget vs. curl situation.

> Since my intended PR was simply adding in a copy of curl.exe and changing the defaults I think I'll hold off on that for now but I can take a look at figuring out how to make those codes use the specified downloader rather than wget explicitly.

I think adding curl.exe, changing the defaults and removing the hardcoded references to wget would work as a first step (I can help out on the latter). Moving all binaries off the main repo would be step 2 in a separate PR.

**Comment by hishamhm** [comment]
Bringing this post to this thread's attention: https://github.com/luarocks/luarocks/issues/273#issuecomment-368772237

**Comment by blandcr** [comment]
Hi @hishamhm 

I finally had a chance to take a look at this today.
* I do think the `download_rocks` function in `spec/util/test_env.lua` should be no problem to adjust to use the configured downloader.
* The `cache.refresh_local_cache` function in `src/luarocks/admin/cache.lua` is a bit trickier since it relies on recursive/mirroring behavior in `wget`. Doing this with `curl` is a bit more involved since it seems we have to download the index then manually invoke `curl` per file and rebuild the tree ourselves locally. I suppose it'd help to know a little more about this cache and how it works/what it does/what it looks like?

**Comment by bluebird75** [comment]
Hi, 
Fixing this issue should be a high priority, either with a new version of wget or switching to curl. I was trying to release a new version of my luaunit, following the luarocks tutorial which advises to put the rock on github release page. I took me more than 1h to realize that the download was not working because of this issue.

This affects potentially any new rock release which followed the official tutorial...

Cheers,

Philippe

**Comment by hishamhm** [comment]
Agreed. I am looking at this. I have a branch with an updated wget.exe that
passes on Appveyor, which I will probably release this weekend, but we are
also looking at replacing with curl.

**Comment by bluebird75** [comment]
Thanks for your support ! Looking forward to the release then:

**Comment by hishamhm** [comment]
Fixed with the release of LuaRocks 2.4.4!

We should still consider dropping the hard-dependency on wget, though!
